src/Libs/GLib-2.0/Internal/SynchronizationContext.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace GLib.Internal;
+
+public sealed class GLibSynchronizationContext : SynchronizationContext
+{
+    #region Fields
+
+    // We need a static variable here instead of a static method because we need
+    // to keep the delegate object to stay alive for the application lifetime. We
+    // don't need a GCHandle for the delegate here because a static is enough to
+    // keep the delegate alive. See https://stackoverflow.com/q/75223488/1829232
+    // for a similar case.
+    private static readonly SourceFunc IdleCallback = data =>
+    {
+        var gch = GCHandle.FromIntPtr(data);
+        var action = (Action)gch.Target!;
+
+        action();
+
+        return false;
+    };
+
+    // According to https://docs.gtk.org/glib/main-loop.html#memory-management-of-sources
+    // we should not free the data in the idle callback because it is stored in a GSource
+    // object, which is still alive when the idle callback is returned. Instead, use
+    // either g_source_remove() or GDestroyNotify.
+    private static readonly DestroyNotify DestroyCallback = data =>
+    {
+        GCHandle.FromIntPtr(data).Free();
+    };
+
+    #endregion
+
+    #region Methods
+
+    public override SynchronizationContext CreateCopy()
+        => new GLibSynchronizationContext();
+
+    public override void Post(SendOrPostCallback d, object? state)
+        => ScheduleAction(() => d(state));
+
+    public override void Send(SendOrPostCallback d, object? state)
+    {
+        using var executed = new ManualResetEvent(false);
+        Exception? exception = null;
+
+        ScheduleAction(() =>
+        {
+            try
+            {
+                d(state);
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+            }
+            finally
+            {
+                executed.Set();
+            }
+        });
+
+        executed.WaitOne();
+
+        if (exception is not null)
+            throw exception;
+    }
+
+    private static void ScheduleAction(Action action)
+    {
+        var data = GCHandle.Alloc(action);
+
+        Internal.Functions.IdleAdd(GLib.Constants.PRIORITY_DEFAULT_IDLE, IdleCallback, GCHandle.ToIntPtr(data), DestroyCallback);
```

Comment:
Aside from this, the `SynchronizationContext` implementation looks good to me

Reply:
I'm not sure if `SourceFuncNotifiedHandler` is really safe from the GC. The moment that `gch.Free()` is called it might be possible for the GC to kick in immediately before we completely return to the GLib, which will cause the thunk for the delegate to be free too. It may seems like a very edge case but it is possible for the time slice of the thread to be used up before we completely return to the GLib, which will cause the OS to swap out our thread to run the other threads.

Reply:
That might be something to investigate for a separate PR and see if a test can reproduce the issue? (maybe from explicitly triggering a GC). 
If there is a fix needed it should be done in the generator (e.g. `CallbackNotifiedHandler.cs` so that all the generated callbacks get fixed

Reply:
@badcel what do you think?

Reply:
- I think calling code in `GLib` from `GLib.Internal` is not a problem as for example enumerations are generated exclusively in the public namespace and are used in the `Internal` namespace.
- I aggree with @cameronwhite. We should use the public code which is available. If there are edge cases we should not optimize here, but fix the problem in the generator. This leads to cleaner code here and fixes the problems in all other places.

Especially the code in the handlers can probably improved quite a lot. Since .NET core there were many improvements in the runtime to avoid allocations. E.g. I suppose that using [UnmanagedCallersOnlyAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute?view=net-6.0) could be a possible solution here. But like @cameronwhite said this is for another PR. I opened #857 to track this separately.

Reply:
`UnmanagedCallersOnlyAttribute` look promising. I'll update the code to use `SourceFuncNotifiedHandler` and fix the format once you have reviewed in a single commit.